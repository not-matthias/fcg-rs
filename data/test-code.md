---
cards-deck: code-test
---

## Testing whether the code works

```rust
fn main() {
    println!("Hello World");
}
```