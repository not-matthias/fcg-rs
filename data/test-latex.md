---
cards-deck: latex-test
---

## This is a question #card

$\top$
