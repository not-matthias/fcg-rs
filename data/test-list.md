---
cards-deck: list-test
---

## This is a question

- Item 1
- Item 2
- Item 3