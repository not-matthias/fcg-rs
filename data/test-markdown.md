## Header 2

Text 1

##### Header 5

Text 2

### Header 3

Text 3

---

# H1

A

## H2

B

## H2
## H2
Content
### H3
Content
### H3
Content
#### H4
Content
#### H4
Content
#### H4
Content

## H2
Content
### H3
Content
#### H4
Content