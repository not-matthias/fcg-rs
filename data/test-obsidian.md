---
cards-deck: obsidian-test
---

# Can this show images? 

Yes:

![[image.png]]
