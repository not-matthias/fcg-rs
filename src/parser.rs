use itertools::Itertools;
use petgraph::graph::NodeIndex;
use petgraph::prelude::EdgeRef;
use petgraph::{Direction, Graph};
use serde::Deserialize;

#[derive(Deserialize, Debug)]
pub struct FileHeader {
    #[serde(rename = "cards-deck")]
    pub cards_deck: String,
}

impl Default for FileHeader {
    fn default() -> Self {
        Self {
            cards_deck: "default".into(),
        }
    }
}

/// The struct representing a markdown header with content.
#[derive(Debug)]
pub struct HeaderWithContent {
    /// The header level.
    pub level: usize,

    /// The header of the block.
    pub header: String,

    /// The content of the header. This will be the back of the card.
    pub content: Vec<String>,
}

pub struct Parser {
    text: String,
}

impl Parser {
    pub fn new(text: String) -> Self {
        Self { text }
    }

    /// Tries to find a parent index for a heading.
    fn find_parent_index(
        graph: &Graph<HeaderWithContent, usize>,
        previous_index: NodeIndex,
        current_level: usize,
    ) -> Option<NodeIndex> {
        // We need to do the following:
        // - Check node at previous_index if it's already at the level above
        // - Go up all parents until we find the

        // Check if the parent is already correct
        //
        let is_parent_correct = |index| {
            if let Some(weight) = graph.node_weight(index) {
                if weight.level < current_level {
                    return Some(index);
                }
            }

            None
        };

        if let Some(index) = is_parent_correct(previous_index) {
            return Some(index);
        }

        // Find the correct parent by going up the chain
        //
        let mut current_index = previous_index;
        while let Some(edge) = graph
            .edges_directed(current_index, Direction::Incoming)
            .last()
        {
            current_index = edge.source();

            if let Some(index) = is_parent_correct(current_index) {
                return Some(index);
            }
        }

        None
    }

    fn parse_markdown(&mut self) -> Graph<HeaderWithContent, usize> {
        let mut graph = Graph::new();
        let mut previous_index: Option<NodeIndex> = None;
        let mut previous_level: Option<usize> = None;

        // Calculate where all the headers start
        //
        let header_indices = {
            let mut header_indices: Vec<_> = self
                .text
                .lines()
                .enumerate()
                .filter_map(|(index, line)| {
                    if line.contains('#') {
                        Some(index)
                    } else {
                        None
                    }
                })
                .collect();

            // We just say that the last line is also a heading so the last tuple has a end.
            //
            header_indices.push(self.text.lines().count());

            header_indices
        };

        // Iterate over the headers
        //
        for (index, next_index) in header_indices.into_iter().tuple_windows() {
            let header = self.text.lines().nth(index).unwrap_or_default();
            let level = header.chars().filter(|c| c.eq(&'#')).count();
            log::trace!("Header {:?} with level {:?}", header, level);

            let content = self
                .text
                .lines()
                .take(next_index)
                .skip(index)
                .filter(|s| !s.is_empty())
                .skip(1) // Skip header
                .collect::<Vec<_>>();
            if content.is_empty() {
                continue;
            }
            log::trace!("Content: {:?}", content);

            // Create the header
            //
            let header = HeaderWithContent {
                level,
                header: header.into(),
                content: content.into_iter().map(|s| s.to_string()).collect(),
            };

            if previous_index.is_some() && previous_level.is_some() {
                // Check if the new header is a sub-header
                // ```
                //       previous
                //     /         \
                //  HEADER     previous
                // ```
                if header.level > previous_level.unwrap() {
                    let node_index = graph.add_node(header);

                    if let Some(parent_index) =
                        Self::find_parent_index(&graph, previous_index.unwrap(), level)
                    {
                        graph.add_edge(parent_index, node_index, level);
                    }

                    previous_index = Some(node_index);
                    previous_level = Some(level);
                }
                // Check if the new header is an upper or equal header
                // ```
                //          HEADER
                //       /         \
                //  previous     previous
                // ```
                else if header.level <= previous_level.unwrap() {
                    let node_index = graph.add_node(header);

                    if let Some(parent_index) =
                        Self::find_parent_index(&graph, previous_index.unwrap(), level)
                    {
                        graph.add_edge(parent_index, node_index, level);
                    }

                    previous_index = Some(node_index);
                    previous_level = Some(level);
                }
            } else {
                previous_index = Some(graph.add_node(header));
                previous_level = Some(level);
            }
        }

        // std::fs::write("out.dot", format!("{:?}", Dot::with_config(&graph, &[])));

        graph
    }

    fn parse_yaml(&mut self) -> FileHeader {
        // The only acceptable input is this:
        // ```
        // ---
        // cards-deck: test
        // ```
        //
        // So we have to find the last `---`:
        //
        let minus_indices: Vec<_> = self
            .text
            .lines()
            .enumerate()
            .filter_map(|(index, line)| {
                if line.contains("---") {
                    Some(index)
                } else {
                    None
                }
            })
            .collect();

        if let Some(last_row) = minus_indices.last() {
            serde_yaml::from_str(&*self.text.lines().take(*last_row).join("\n")).unwrap_or_default()
        } else {
            Default::default()
        }
    }

    /// Parses and returns the header and markdown content as graph.
    pub fn parse(&mut self) -> (FileHeader, Graph<HeaderWithContent, usize>) {
        (self.parse_yaml(), self.parse_markdown())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_markdown() {
        let content = include_str!("../data/test-markdown.md").to_string();
        let mut parser = Parser::new(content);
        let content = parser.parse_markdown();

        insta::assert_debug_snapshot!(content);
    }

    #[test]
    fn parse_yaml() {
        let content = "---\ncards-deck: deck-name\n---".to_string();
        let mut parser = Parser::new(content);
        let content = parser.parse_yaml();

        insta::assert_debug_snapshot!(content);
    }
}
