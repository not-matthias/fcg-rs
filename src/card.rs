use crate::parser::HeaderWithContent;
use itertools::Itertools;
use petgraph::data::DataMap;
use petgraph::graph::NodeIndex;
use petgraph::prelude::EdgeRef;
use petgraph::{Direction, Graph};
use std::borrow::Cow;

#[derive(Debug)]
pub struct Card {
    pub(crate) front: String,
    pub(crate) back: String,
}

impl Card {
    fn header_context(graph: &Graph<HeaderWithContent, usize>, index: NodeIndex) -> Vec<String> {
        let mut context_data = Vec::new();
        let mut current_index = index;

        while let Some(edge) = graph
            .edges_directed(current_index, Direction::Incoming)
            .next()
        {
            current_index = edge.source();

            if let Some(context) = graph.node_weight(current_index).map(|h| h.header.clone()) {
                context_data.push(context.into());
            }
        }

        // We went from bottom up, so we have to reverse the vec to get the correct order.
        //
        context_data.reverse();

        context_data
    }

    pub fn new(graph: &Graph<HeaderWithContent, usize>, index: NodeIndex) -> Self {
        let node = graph.node_weight(index).unwrap();
        let context = Self::header_context(graph, index);

        // Remove the unneeded `#` from the context
        //
        let context = context
            .into_iter()
            .map(|c| c.replace("# ", "").replace("#", ""))
            .join(" > ");

        let back = node.content.join("\n").trim().to_string();
        let back = comrak::markdown_to_html(&back, &comrak::ComrakOptions::default());

        Self {
            front: format!("{} > {}", context, node.header.replace('#', "").trim()).into(),
            back: back,
        }
    }
}
