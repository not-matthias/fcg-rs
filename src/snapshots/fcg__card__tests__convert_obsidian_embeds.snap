---
source: src/card.rs
expression: converted
---
---
cards-deck: obsidian-test
---

# Can this show images? 

Yes:

<img src='data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAa8AAAEzCAYAAACGz4w5AAARCElEQVR4Ae3dW24jNxAFUDNf2v9eB/qcwHFs+CFLzRbJKpInwIwtqZusOqXgeqxX+fPnz98X/6UWKKW8/P1rTKmHpDgCBIYK/DN0N5udErhcLqfOcxIBAgRWFRBeE0z2er1OUOUTJTqVAAEClQLCqxLM4QQIECAQLyC84megAgIE4gVUMJmA8JpsYMolQIAAgZcX4bXhvaCUsmHXWiZAYCWBJcNrpQH16MWzF3uoWpMAgZECwmukdpK9PHsxySCUQYDAaQHhdZrOiQTuCbiNAIGeAsKrp661CRAgQKCLgPDqwmpRAgQIxAusXIHwWnm6eiNAgMCiAsJr0cFqiwABAisLCK9ZpqtOAgQIEPgQEF4fFL4hQIAAgVkEhNcsk1IngXgBFRBIIyC80oxCIQQIECBwVEB4HZVyHAECBAjEC/xfgfD6H8IXAgQIEJhHQHjNMyuVEiBA4KUUnwrxejcQXq8K/gQJ2JYAgVoBnwrxJia83hz8TYAAgSkEfCrE25iE15uDvwkQ2FRA23MKCK8556ZqAgQIbC0gvLYev+YJECAwp8Ba4TXnDFRNgAABApUCwqsSzOEECBAgEC8gvOJnoIK1BHRDgMAAAeE1ANkWBAgQINBWQHi19bQageYCpXhHheaoqy+4QX/Ca4Mha3FuAe+oMPf8jlZfih9Sjlq9Hie8XhX8IZBYwDsqJB5Ow9L8kFKHKbzqvAKOtiUBAjsI+CGlbsrCq87L0QQIECCQQEB4JRiCEghkF1AfgWwCW4dXKR4gzXaHVA8BAgSOCGwdXh4gPXIXcQwBAgQyCHytYevw8gDp1zuDSwQIEPgsUEre305tHV6fh+R7AgQIEPgqkPm3U8Lr66xcGiNgFwIEJhDI/Nsp4TXBHUiJBAgQIPBVQHh99XCJAIFdBPQ5tYDwmnp8iidAgMCeAsJrz7nrmgABAlMLLBJeU89A8QQIECBQKSC8KsEcToAAAQLxAsIrfgYqWERAGwQIjBMQXuOs7USAAAECjQSEVyNIyxAgQCBeYJ8KhNc+s9YpAQIElhEQXsuMUiMECBDYR0B45Z21yggQIEDgFwHh9QvMiKtLyftxAyP6twcBAgTOCgivs3INzsv8cQMN2rPECgJ6IJBUQHgFDibzxw0EstiaAAECDwWE10MiBxAgQIBAoMDNrYXXTRZXEiBAgEBmAeGVeTpqI0CAAIGbAsLrJosrewlYlwABAi0EhFcLRWsQIECAwBCBUt5eYiS8hnDbhACBPAIqmVng/SVGwmvmKaqdAAECmwm8v8RIeG02eO0SIEBgBYHZw2uFGeiBAAECBCoFhFclmMMJECBAIF5gm/Aq5e0ZKvHkKlhOQEMECAwX2Ca83p+hMlzYhgQIECDQXGCb8Hp/hkpzQQsSIEAgXmC7CrYJr+0mq2ECBAgsLCC8Fh6u1ggQILCqgPBKOFklESBAgMB9AeF138etBAgQIJBQQHglHIqSCMQLqIBAbgHhlXs+qiNAgACBGwLC6waKqwgQIEAgXuBeBcLrno7bCBAgQCClgPBKORZFESBAgMA9AeF1T8dt7QSsRIAAgYYCwqshpqUIECBAYIyA8BrjbBcCBOIFVLCQgPBaaJhaIUCAwC4CwmuXSeuTAAECCwlMG14LzUArBAgQIFApILwqwRxOgAABAvECwit+BiqYVkDhBAhECQivKHn7EggQKKUE7GpLAu0FhFd7UysSSCtwuVzS1qawcwK7niW8dp28vrcUuF6vW/at6fUEhNd6M9XRogKl+JXfoqPV1gkB4XUCrdspFiZwR8Cv/O7guGk7AeG13cg1PKuAX/nNOjl19xAQXj1UrUlgXgGVE5hCQHhNMSZFEiBAgMBnAeH1WePG96V4kPwGi6sIECDQT+DAysLrAZIHyR8AuZkAAQIBAsLrAboHyR8AuZkAAQIBAsIrAH2vLXVLgACB9gLCq72pFQkQIECgs4Dw6gxseQIE4gVUsJ6A8FpvpjoiQIDA8gLCa/kRa5AAAQLrCcwXXuvNQEcECBAgUCkgvCrBHE6AAAEC8QLCK34GKphPQMUECAQLCK/gAdieAAECBOoFhFe9mTMIECAQL7B5BcJr8zuA9gkQIDCjgPCacWpqJkCAwOYCwivFHUARBAgQIFAjILxqtBxLgAABAs0FSqn/3ETh1XwMFiQwp4CqCUQJnPncROEVNS37EiBAgMB/Amc+N1F4/UfnLwIECBCIFzhegfA6buVIAgQIEEgiILySDEIZBAgQIHBcQHgdt3JknYCjCRAg0E1AeHWjtTABAgQI9BIQXr1krUuAQLyACpYVEF7LjlZjBAgQWFdAeK07W50RIEBgWYGJwmvZGWiMAAECBCoFhFclmMMJECBAIF5AeMXPQAUTCSiVAIEcAsIrxxxUQYAAAQIVAsKrAsuhBAgQiBdQwauA8HpV8IcAAQIEphIQXlONS7EECBAg8CogvF4V4v7YmQABAgROCAivE2hOIUCAAIFYAeEV6293AvECKiAwoYDwmnBoSiZAgMDuAsJr93uA/gkQIBAvUF2B8KomcwIBAgQIRAsIr+gJ2J8AAQIEqgWEVzWZEx4JuJ0AAQK9BYRXb2HrEyBAgEBzAeHVnNSCBAjEC6hgdQHhtfqE9UeAAIEFBYTXgkPVEgECBFYXmCG8Vp+B/ggQIECgUkB4VYI5nAABAgTiBYRX/AxUMIOAGgkQSCUgvFKNQzEECBAgcERAeB1RcgwBAgTiBVTwSUB4fcLwLQECBAjMISC85piTKgkQIEDgk4Dw+oQx8lt7ESBAgMB5AeF13s6ZBAgQIBAkILyC4FfdtpSyamsL9qUlAvMKCK95Z5ey8svlkrIuRREgsJaA8FprnuHdXK/X8BoUQIDAPAJnKxVeZ+WcR4AAAQJhAsIrjN7GBAgQIHBW4Ed4leIB97OY258HgAABAoMEfoSXB9wHyduGAAECBE4L/AgvD7iftnQiAQLxAirYROBHeG3StzYJECBAYGIB4TXx8JROgACBXQVSh9euQ9H3MYFSPLnomJSjCKwnILzWm+k2HXly0Taj1iiBHwLC6weJK2YRGPPkolk01ElgLwHhtde8dUuAAIElBITXEmPUBAECKwvo7aeA8Ppp4hoCBAgQSC4gvJIPSHkECBAg8FNAeP006XuN1QncESjF0//v8LiJwIeA8Pqg8A2BeAFP/4+fwTMVlOKHj2f8as4VXjVajiXQWWDQ0/87d7Hv8n74GDd74TXO2k4ECCwu4IePcQMWXuOs7USAAAEC7wJPfhVeTwI6nQABAgTGCwiv8eZ2JECAAIEnBYTXk4BOfxXwhwABAmMFhNdYb7sRIECAQAMB4dUA0RIECMQLqGAvAeG117x1S4AAgSUEhNcSY9QEAQIE9hLIGV57zUC3BAgQIFApILwqwRxOgAABAvECwit+BirIKaAqAgQSCwivxMNRGgEC8QKleKf4+Cn8rEB4/TRxDQECBD4EQt8p/qMK33wXEF7fRVwmQIDAJwHvFP8JI9G3wivRMJRCgAABAscEhNcxpwZHWYIAAQIEWgkIr1aS1iFAgACBYQLCaxi1jQjEC6iAwCoCwmuVSeqDAAECGwkIr42GrVUCBAjEC7SpQHi1cbQKAQIECAwUEF4DsW1FgAABAm0EhFcbx11X0TcBAgRCBIRXCLtNCRAgQOAZAeH1jJ5zCRCIF1DBlgLCa8uxa5oAAQJzCwivueenegIECGwp0Dy8Snnms2+2nIGmCRAgQKBSoHl4+eybygk4nAABAgSqBZqHl8++qZ6BE5IJKIcAgfwCzcMrf8sqJECAAIHZBYTX7BNUPwECCwpo6ZGA8Hok5HYCBAgQSCcgvNKNREEECBAg8EhAeD0Sev52KxAgQIBAYwHh1RjUcgQIECDQX0B49Te2A4F4ARUQWExAeC02UO0QIEBgBwHhtcOU9UiAwGmBUrzl3Wm8ryc2vSS8mnJajACB1QS85V3OiQqvnHNRFQECSQS85V2SQXwrQ3h9A3HxmICjCBAgECkgvCL17U2AAAECpwSE1yk2JxEgEC+ggp0FhNfO09c7AQIEJhUQXpMOTtkECBDYWSBLeO08A70TIECAQKWA8KoEczgBAgQIxAsIr/gZqCCLgDoIEJhGQHhNMyqFEiBAgMC7gPB6l/CVAAEC8QIqOCggvA5COYwAAQIE8ggIrzyzUAkBAgQIHBQQXgehzhzmHAIECBDoIyC8+rhalQABAgQ6CgivjriWJhAvoAICawoIrzXnqisCBAgsLSC8lh6v5ggQIBAv0KOC6cKrlNLDwZoECBAgMJHAdOF1uVwm4lUqAQIECPQQmC68rtdrDwdrHhVwHAECBBIITBdeCcyUQIAAAQLBAqfDqxSPPQXPzvbJBUrx/0inEVmWwMvp8PLYk3sPgfsC/h+57+NWAs8IlD9//vx9ZgHnEiBAgACB0QKn/+XVqlDrECBAgACBWgHhVSvmeAIECBAIFxBe4SNQQLyACggQmE1AeM02MfUSIECAwPlnG7IjQIAAgXYCVqoT8C+vOi9HEyBAgEACAeGVYAhKIECAAIE6AeFV53XsaEcRIECAQFcB4dWV1+IECBAg0ENAePVQtSaBeAEVEFhaQHgtPd51myvFm96uO12dEXgsILweGzkioYA3vU04FCUR+C7Q8bLw6ohr6X4C1+u13+JWJkAgvYDwSj8iBRIgQIDAdwHh9V3E5V8EXE2AAIE8AsIrzyxUQoAAAQIHBYTXQSiHESAQL6ACAu8CwutdwlcCBAgQmEZAeE0zKoUSIECAwLtAXHi9V+ArAQIEHgiU4kXpD4i2u1l4bTdyDROYT8CL0uebWe+KhVdvYetnFlDbJAJelD7JoAaWKbwGYtuKAAECBNoICK82jlYhQIDAOQFnnRIQXqfYnETgd4FSPLngdx23EGgjILzaOFqFwIeAJxd8UPiGQDcB4dWU1mIEXl48ucC9gEB/AeHV39gOBAgQINBYQHg1BrUcgWgB+x8XKMXjk8e1ch0pvHLNQzUECAwU8PjkQOzGWwmvxqCWI0BgHgGPT/aaVf91hVd/YzsQIECAQGMB4dUY1HIECBAg0F9AePU3nn0H9RMgQCCdgPBKNxIFESBAgMAjAeH1SMjtBAjEC6iAwDcB4fUNxEUCBAgQyC8gvPLPSIUECBAg8E0gILy+VbDBxVK8in+DMWuRAIGBAsJrALZX8Q9AtgUBAlsJCK8B4/Yq/gHIlVs4nACBuQWE19zzUz0BAgS2FBBeW459jaZL8VjiGpPctQt9PyMgvJ7Rc26ogMcSQ/ltTiBU4HR4leKn3tDJ2dwnFrsPENhY4HR4+an3y73GBQIECBAYKHA6vDyDbuCUbEWAAAECXwROh9eXVVwgQCBeQAUENhIQXhMNuxSPM040LqUSINBRQHh1xG29tMcZW4tajwCBxgLDlhNew6if38jjjM8bWoEAgTUEhNcac9QFAQIEthIQXluNu65ZRxMgQCCrgPDKOhl1ESBAgMCvAsLrVxo3ECAQL6ACArcFhNdtF9cSIECAQGIB4ZV4OEojQIAAgdsCI8PrdgWuJUCAAAEClQLCqxLM4QQIECAQLyC84meggpEC9iJAYAkB4bXEGDVBgACBvQSE117z1i0BAvECKmggILwaIFqCAAECBMYKCK+x3nYjQIAAgQYCwutJRKcTIECAwHgB4TXe3I4EwgVK8cGm4UNQwFMCwuspPicTyCBQX8Plcqk/yRkEEgkIr0TDUAqBUQI+2HSUtH16CQivXrLWJUCAwEYCo1sVXqPF7UeAAAECTwsIr6cJLUCAAAECowWE12jxGfZTIwECBJIL/AvQtW1UWmiWUQAAAABJRU5ErkJggg'>

